### PR TITLE
feat(ena-submission): Upload ena submission results to Loculus

### PR DESCRIPTION
**Pull request by anna-parker**
resolves https://github.com/loculus-project/loculus/issues/2424

preview URL: https://upload-ena-submission-res.loculus.org/

### Summary
1. This PR encompasses the entire ena-submission process, for details see: [ena-submission/README.md](https://github.com/loculus-project/loculus/pull/2417/files#diff-fbb6131a0af34856a7dfa00148392503ea6c10bcb83279c948c4dab00323e2d3).

2. It adds database connection pooling, otherwise I get this error:
image
caused by Exhaustion of Local Ports: "When your Python application makes connections to the database, it uses a local (ephemeral) port on your machine. If your application opens many connections in a short period and doesn’t properly close them, you might run out of available local ports." - it might be possible to fix this just with better closing of port connections... I checked and the postgresDB can support `SHOW max_connections;` = 100. After the ena-submission-pod jobs were killed there were still `SELECT count(*) FROM pg_stat_activity;` = 20 connections to the db. Probably from the backend. I wasn't closing db updates (this would explain ~20-30 additional connections).

3. It also fixes an issue in the backend when merging external metadata with other metadata fields, an alternative solution could be: 

```sql
create or replace function jsonb_concat(a jsonb, b jsonb) 
returns jsonb
language plpgsql
immutable
parallel safe
as $$
declare
    result jsonb;
begin
    -- Merge the two JSON objects
    result := a || jsonb_object_agg(key, 
                case 
                    when a->key is null or a->key = 'null' then b->key 
                    else a->key 
                end)
                from jsonb_each_text(b) as elem(key, value);
    return result;
end;
$$;
```
and use this like:
```sql
case 
       when all_external_metadata.external_metadata is null then jsonb_build_object('metadata', (sequence_entries_preprocessed_data.processed_data->'metadata'))
        else jsonb_build_object('metadata', (sequence_entries_preprocessed_data.processed_data->'metadata') || all_external_metadata.external_metadata)
    end as joint_metadata
```

### PR Checklist

- [x] Confirm this works locally from uploading new sequences, getting the sequences from the slack notification and uploading to the github repo to appearing on the page of the uploaded sequence: 
image
Note Biosamples was down so I had to artificially add the values in.
- [x] Confirm this works on a preview instance: test full cycle from uploading sequences to the preview instance. Similar to in https://github.com/loculus-project/loculus/pull/2332 I had to log into the database via port forwarding to set the sample results to existing entries: 
```
{
        "ena_sample_accession":  "ERS20170050",
        "biosample_accession": "SAMEA115670243",
        "ena_submission_accession": "ena_submission_accession"
    }
````
(as biosamples is currently down), and then did the same for the erz_accession (this was successful but to not have to wait I set this to an existing erz accession). This finishes the submission cycle and then the results are sent to loculus: 
image
and are seen on the website (with links that work except for sra... it seems like this isn't actually an sra accession??):

image
now without sra_run_accession: image
- [x] Check this works for a single segmented organism

### CODE REVIEW

File: ena-submission/scripts/create_project.py
Comment (chaoran-chen) [inline ena-submission/scripts/create_project.py:249]: Is a SQL injection possible through the `institutions` field? Because it is passed on into `update_db_where_conditions` which uses string concatenation to construct the SQL.

**Review by chaoran-chen — DISMISSED** [review]
I think that there is at least one possible way for an SQL injection.

There are multiple places where a function constructs an SQL query via string concatenation using values passed into the function without escaping and validation. This is very dangerous and even for places where an attack is currently not possible because the input values are all under our control (e.g., the `organism` field), this could potentially lead to future vulnerabilities.

File: ena-submission/scripts/call_loculus.py
Comment (chaoran-chen) [inline ena-submission/scripts/call_loculus.py:97]: Is it intentional to not use `response.ok`?

File: ena-submission/scripts/call_loculus.py
Comment (chaoran-chen) [inline ena-submission/scripts/call_loculus.py:129]: Huh? Why this change? I thought that `make_request` will already raise an error if the response is not ok.

File: ena-submission/scripts/create_assembly.py
Comment (chaoran-chen) [inline ena-submission/scripts/create_assembly.py:517]: Would it make sense to sleep a bit between the iterations?

File: ena-submission/scripts/ena_submission_helper.py
Comment (chaoran-chen) [inline ena-submission/scripts/ena_submission_helper.py:261]: Sorry for bringing this up if it has been discussed already and it was decided to be the right way. I'm just a bit surprised to see a gap being replaced by an `N`. Wouldn't it make more sense to simply remove the `-` as they represent emptiness?

File: ena-submission/scripts/ena_submission_helper.py
Comment (chaoran-chen) [inline ena-submission/scripts/ena_submission_helper.py:294]: Is this safe? It looks like a potential source for an injection. Should we perform some escaping or validations (e.g., check that user-specified fields such as platform/sequencingProtocol do not contain forbidden/problematic characters)?

File: ena-submission/scripts/create_assembly.py
Comment (chaoran-chen) [inline ena-submission/scripts/create_assembly.py:108]: I think the result type is wrong, this function is returning a `AssemblyManifest`, isn't it?

File: ena-submission/scripts/ena_submission_helper.py
Comment (theosanderson) [reply ena-submission/scripts/ena_submission_helper.py:261]: Or to handle this outside ENA submission - IMO Loculus should never contain raw sequences with `-`s

**Comment by chaoran-chen** [comment]
Thanks @anna-parker for this huge piece of work!! I read (rather briefly) through the code and commented whenever I noticed something. I found a few vulnerabilities that should be fixed. Otherwise, it looks great for now!

File: backend/src/main/resources/db/migration/V1.2__update_view.sql
Comment (theosanderson) [inline backend/src/main/resources/db/migration/V1.2__update_view.sql:1]:
```suggestion
drop view if exists external_metadata_view cascade;
```

File: ena-submission/README.md
Comment (theosanderson) [inline ena-submission/README.md:21]:
```suggestion
In our kubernetes pod we run flyway in a docker container, however when running locally it is best to [download the flyway CLI](https://documentation.red-gate.com/fd/command-line-184127404.html) (or `brew install flyway` on macOS).
```

File: ena-submission/README.md
Comment (theosanderson) [inline ena-submission/README.md:84]:
```suggestion
It is then the reviewer's turn to review these sequences. [TODO: define review criteria] If these sequences meet our criteria they should be uploaded to [pathoplexus/ena-submission](https://github.com/pathoplexus/ena-submission/blob/main/approved/approved_ena_submission_list.json) (currently we read data from the [test folder](https://github.com/pathoplexus/ena-submission/blob/main/test/approved_ena_submission_list.json) - but this will be changed to the `approved` folder in production). The `trigger_submission_to_ena` rule is constantly checking this folder for new sequences and adding them to the submission_table if they are not already there. Note we cannot yet handle revisions so these should not be added to the approved list [TODO: do not allow submission of revised sequences in `trigger_submission_to_ena`]- revisions will still have to be performed manually.
```

Comment (theosanderson) [reply ena-submission/README.md:84]: Can we raise an Issue for the second TODO before merging this

File: ena-submission/scripts/create_project.py
Comment (anna-parker) [reply ena-submission/scripts/create_project.py:249]: I have modified the function to now always use `%s` for passing inputs and validating the name.

File: ena-submission/scripts/create_project.py
Comment (anna-parker) [reply ena-submission/scripts/create_project.py:294]: I think this should now be resolved with the `%s`

File: ena-submission/scripts/submission_db_helper.py
Comment (chaoran-chen) [inline ena-submission/scripts/submission_db_helper.py:142]: I think that this implementation is already entirely safe for now, but to make the function even more future-proof, it would be good to validate the keys as well. A simple implementation, where you wouldn't need to list all possible keys, and that should be enough for this case, could be to check that the key only consists of letters (A-Za-z), numbers, and underscores.

File: ena-submission/scripts/submission_db_helper.py
Comment (chaoran-chen) [inline ena-submission/scripts/submission_db_helper.py:168]: To improve readability, the constants such as `"HAS_ERRORS"` can be moved directly into the SQL. Otherwise, this looks great and safe!

File: ena-submission/ENA_submission.md
Comment (anna-parker) [reply ena-submission/ENA_submission.md:38]: I have updated with current state

File: ena-submission/scripts/ena_submission_helper.py
Comment (anna-parker) [reply ena-submission/scripts/ena_submission_helper.py:261]: I removed '-' from the example data: https://github.com/pathoplexus/example_data/pull/4 - I will add a prepro check before removing this code.

File: ena-submission/scripts/call_loculus.py
Comment (anna-parker) [reply ena-submission/scripts/call_loculus.py:96]: Yes we can just use that - I changed now to be more consistent

File: ena-submission/README.md
Comment (anna-parker) [reply ena-submission/README.md]: I added a check for this in the code to prevent us from accidentally submitting revisions :-) there is already an issue on figuring out how to handle revisions to ENA: https://github.com/loculus-project/loculus/issues/2428

File: ena-submission/scripts/submission_db_helper.py
Comment (anna-parker) [reply ena-submission/scripts/submission_db_helper.py:161]: I added a function to perform an explicit check that all keys in the conditions dictionary are actual columns.

File: ena-submission/scripts/ena_submission_helper.py
Comment (anna-parker) [reply ena-submission/scripts/ena_submission_helper.py:294]: I am not quite sure what you mean - this function produces a manifest file and isn't added to the postgres database. The file is sent to ENA using their webin-cli so I don't think there is any injection risk here? But maybe I don't understand?

**Review by corneliusroemer — APPROVED** [review]
I'll review as we go - let's try this out

File: ena-submission/scripts/ena_submission_helper.py
Comment (anna-parker) [reply ena-submission/scripts/ena_submission_helper.py:293]: created a follow up issue: https://github.com/loculus-project/loculus/issues/2831

File: ena-submission/scripts/create_assembly.py
Comment (anna-parker) [reply ena-submission/scripts/create_assembly.py:521]: created a follow up issue: https://github.com/loculus-project/loculus/issues/2831

File: ena-submission/scripts/create_project.py
Comment (anna-parker) [reply ena-submission/scripts/create_project.py:68]: created a follow up issue: https://github.com/loculus-project/loculus/issues/2831

**Comment by anna-parker** [comment]
@chaoran-chen sorry I had to dismiss your review to merge this (but I did address and fix the issues you found), from discussion with @rneher and @corneliusroemer we will merge this now and work on https://github.com/loculus-project/loculus/issues/2832 in a separate PR.

**Pull request merged**
